test/utils/lazy.test.ts

```diff
@@ -0,0 +1,117 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See LICENSE.md in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as assert from 'assert';
+import { Lazy, AsyncLazy } from "../../extension.bundle";
+import { delay } from '../../extension.bundle';
+
+suite('(unit) Lazy tests', () => {
+    suite('Lazy<T>', () => {
+        test('Normal', async () => {
+            let factoryCallCount = 0;
+            const lazy: Lazy<boolean> = new Lazy(() => {
+                factoryCallCount++;
+                return true;
+            });
+
+            lazy.value;
+            lazy.value;
+
+            assert.equal(factoryCallCount, 1, 'Incorrect number of value factory calls.');
+        });
+
+        test('With lifetime', async () => {
+            let factoryCallCount = 0;
+            const lazy: Lazy<boolean> = new Lazy(() => {
+                factoryCallCount++;
+                return true;
+            }, 5);
+
+            lazy.value;
+            lazy.value;
+
+            assert.equal(factoryCallCount, 1, 'Incorrect number of value factory calls.');
+
+            await delay(10);

```

Comment:
It is kind of nice that Node is, to a large extent, deterministic in terms of how callbacks are queued and executed

Reply:
One thing I'm not sure about is, if you have two calls waiting on the same promise--first A then B--I know A starts then B starts. But, does A finish and then B finish, or the other way around, or random-ish?